packages/embark/src/lib/modules/ens/index.js

```diff
@@ -558,17 +558,19 @@ class ENS {
     ], cb);
   }
 
-  isENSName(name, callback = () => {}) {
+  isENSName(name, callback) {
```

Comment:
Why change this, did it cause an issue?

Reply:
Actually, this one should be changed further, I forgot to go back and do that.

I think it should either be guaranteed synchronous:

```js
isENSName(name) {
  if (typeof name !== 'string') {
    return false;
  }
  return Boolean(ENS_WHITELIST.find(ensExt => name.endsWith(ensExt)));
}
```

Or it should be guaranteed async:

```js
isENSName(name, callback) {
  const isENSName = async (name) => {
    if (typeof name !== 'string') {
      return false;
    }
    return Boolean(ENS_WHITELIST.find(ensExt => name.endsWith(ensExt)));
  }

  if (callback) {
    return callbackify(isENSName)(name, callback);
  }
  return isENSName(name);
}
```

Ambiguity in control flow at a callsite is nearly always a bad thing. `callbackify` ensures a callback is fired on `nextTick`, so in the second version `isENSName` will always involve async control flow.

Does that make sense? Should I make it consistently async or do you think it's okay if it's always sync. All the places it's used seem to not supply a callback, but I may be missing something.

Reply:
So, I thought about this a bit more and I realize my explanation is bit off, my apologies.

There's nothing inherently wrong with synchronous execution of callbacks. `Array#forEach`, `Array#map` and many others use them.  The important thing is that callback control flow is *consistent*, i.e. you don't want it called synchronously sometimes and asynchronously other times.

Also, given `function foo(a, b, (err, val) => {...})`, if `foo` has that signature in a NodeJS context then this is pretty much expected:

```
foo(a, b, (err, val) => {
  // A Code -- runs after B Code
})
// B Code -- runs before A Code
```

The reason being, of course, that `foo` generally does something asynchronous before invoking the callback.

Anyway, back to `isENSName`. It calls its callback synchronously but looking at its signature my first instinct is that it will run like `foo` in the example, if it's invoked with a callback.

So, I still suggest revising it one way or the other as in my previous comment, but I'd also suggest revising the callback signature to be `(err, value)` if we make it consistently async.

Reply:
I actually don't remember why we put it with a callback, since it clearly could just be sync.
Anyway, do what you think is best, but I think getting rid of the callback is probably simplest.

Reply:
Done. See: d8a72ff624631d34ac9cd3d052f97ba96b02f5df.

Reply:
Update: b72b171b209c6f895e1087f3a009b2cfa700e4ee.

The `TODO` comment I left there represents a broad principle for refactoring many of embark's modules over time:

```js
// TODO in general (not just this function) control flow at callsite of
// internal/external APIs should be consistently async or sync:
//   + shouldn't both return a meaningful value and invoke callback, should
//     yield control flow to caller / it's callback one way or the other
//   + should return a promise if no callback is supplied
//   + NodeJS style callback-last callbacks should be expected to have
//     consistent `(err, value)` signature/behavior
//   + callback should executed async; see: built-in util.callbackify
```

For now, I think we can proceed with this PR as-is.